hpx/util/detail/unwrap_impl.hpp

```diff
@@ -0,0 +1,294 @@
+//  Copyright (c) 2017 Denis Blank
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_DETAIL_UNWRAP_IMPL_HPP
+#define HPX_UTIL_DETAIL_UNWRAP_IMPL_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/traits/future_traits.hpp>
+#include <hpx/traits/is_future.hpp>
+#include <hpx/traits/is_tuple_like.hpp>
+#include <hpx/util/invoke.hpp>
+#include <hpx/util/invoke_fused.hpp>
+#include <hpx/util/pack_traversal.hpp>
+#include <hpx/util/result_of.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace hpx {
+namespace util {
+    namespace detail {
+        /// This struct defines a configurateable environment in order
+        /// to adapt the unwrapping behaviour.
+        ///
+        /// The old unwrap (unwrapped) allowed futures to be instanted
+        /// with void whereas the new one doesn't.
+        ///
+        /// The old unwrap (unwrapped) also unwraps the first occuring future
+        /// independently a single value or multiple ones were passed to it.
+        /// This behaviour is inconsistent and was removed in the
+        /// new implementation.
+        ///
+        /// \note If the old implementation of unwrapped gets removed all
+        ///       the configuration mechanis here can be removed while
+        ///       preserving the behaviour of the new implementations
+        ///       configuration.
+        ///
+        template <bool AllowsVoidFutures, bool UnwrapTopLevelTuples>
+        struct unwrap_config
```

Comment:
@hkaiser Ah the actual issue here is that the code is used from both versions: the new and the old one. Thus I decided to use compile-time value based configuration here intead of preprocessor macros.
However, this way makes it difficult to remove the real old unwrapped behaviour since it will be still available in the dead paths of the configuration then.

Maybe it would be better here, to just duplicate the file into two pieces, where one can be removed when the old unwrapped is removed (this is also better for the compile time).